File: tests/algorithms/prime_sieve/prime_sieve.py
#!/usr/bin/env python3
"""
Prime sieve benchmark implementation in Python.
Finds all prime numbers up to n using the Sieve of Eratosthenes.
"""

import sys
import time
import math


def sieve_of_eratosthenes(n):
    """Find all prime numbers up to n using the Sieve of Eratosthenes."""
    if n < 2:
        return []
    
    is_prime = [True] * (n + 1)
    is_prime[0] = is_prime[1] = False
    
    for i in range(2, int(math.sqrt(n)) + 1):
        if is_prime[i]:
            for j in range(i * i, n + 1, i):
                is_prime[j] = False
    
    return [i for i in range(2, n + 1) if is_prime[i]]


def main():
    """Main execution function."""
    n = 100000
    
    print(f"Finding all primes up to {n}...")
    start_time = time.perf_counter()
    
    primes = sieve_of_eratosthenes(n)
    
    end_time = time.perf_counter()
    execution_time = end_time - start_time
    
    print(f"Result: Found {len(primes)} primes")
    print(f"Largest prime: {max(primes) if primes else 'None'}")
    print(f"Execution time: {execution_time:.6f} seconds")


if __name__ == "__main__":
    main()

File: tests/algorithms/prime_sieve/prime_sieve.ts
#!/usr/bin/env node
/**
 * Prime sieve benchmark implementation in TypeScript.
 * Finds all prime numbers up to n using the Sieve of Eratosthenes.
 */

function sieveOfEratosthenes(n: number): number[] {
    if (n < 2) {
        return [];
    }
    
    const isPrime = new Array(n + 1).fill(true);
    isPrime[0] = isPrime[1] = false;
    
    const limit = Math.floor(Math.sqrt(n));
    
    for (let i = 2; i <= limit; i++) {
        if (isPrime[i]) {
            for (let j = i * i; j <= n; j += i) {
                isPrime[j] = false;
            }
        }
    }
    
    const primes: number[] = [];
    for (let i = 2; i <= n; i++) {
        if (isPrime[i]) {
            primes.push(i);
        }
    }
    
    return primes;
}

function main(): void {
    const n = 100000;
    
    console.log(`Finding all primes up to ${n}...`);
    const start = performance.now();
    
    const primes = sieveOfEratosthenes(n);
    
    const end = performance.now();
    const executionTime = (end - start) / 1000; // Convert to seconds
    
    console.log(`Result: Found ${primes.length} primes`);
    if (primes.length > 0) {
        console.log(`Largest prime: ${primes[primes.length - 1]}`);
    }
    console.log(`Execution time: ${executionTime.toFixed(6)} seconds`);
}

if (require.main === module) {
    main();
}